File: core/src/main/java/org/github/krandom/generator/base/ShortGenerator.java
/*
 * Copyright (c) 2026 krandom contributors
 *
 * Licensed under the MIT License. See LICENSE in the project root for license information.
 */
package org.github.krandom.generator.base;

/**
 * Generates random {@link Short} values.
 *
 * <p>Default range: [{@code Short.MIN_VALUE}, {@code Short.MAX_VALUE}) = [-32768, 32767).
 *
 * <pre>{@code
 *   short value   = new ShortGenerator().generate();
 *   short inRange = new ShortGenerator((short) 100, (short) 200).generate();
 * }</pre>
 */
public final class ShortGenerator extends AbstractBoundedGenerator<Short> {

    public ShortGenerator() {
        super(Short.MIN_VALUE, Short.MAX_VALUE, null);
    }

    public ShortGenerator(short min, short max) {
        super(min, max, null);
    }

    public ShortGenerator(short min, short max, long seed) {
        super(min, max, seed);
    }

    /**
     * Generate a short in the half-open range [{@code min}, {@code max}).
     *
     * @throws IllegalArgumentException if {@code min == max}
     */
    @Override
    public Short generate(Short min, Short max) {
        validate(min, max);
        int lo = lo(min, max).intValue();
        int hi = hi(min, max).intValue();
        return (short) random.nextInt(lo, hi);
    }
}

File: core/src/main/kotlin/org/github/krandom/common/generators/ShortGenerator.kt
/*
 * Copyright (c) 2026 krandom contributors
 *
 * Licensed under the MIT License. See LICENSE in the project root for license information.
 */
package org.github.krandom.common.generators

import org.github.krandom.common.RangedTypeGenerator
import java.security.SecureRandom
import java.util.Random

/**
 * Generates random [Short] values.
 *
 * [generate(start, end)] uses a half-open range [start, end): start is inclusive, end is exclusive.
 */
class ShortGenerator(
    private val start: Short = Short.MIN_VALUE,
    private val end: Short = Short.MAX_VALUE,
    seed: Long? = null
) : RangedTypeGenerator<Short> {

    private val random: Random = if (seed != null) Random(seed) else SecureRandom()

    override fun generate(): Short = generate(start, end)

    override fun generate(start: Short, end: Short): Short {
        require(start != end) { "start ($start) and end ($end) must differ" }
        val lo = minOf(start, end).toInt()
        val hi = maxOf(start, end).toInt()
        return random.ints(1, lo, hi).findFirst().orElse(lo).toShort()
    }
}
